pystac_client/collection_search.py

```diff
@@ -0,0 +1,505 @@
+            stac_io=stac_io,
+            client=client,
+            limit=limit,
+            bbox=bbox,
+            datetime=datetime,
+            query=query,
+            filter=filter,
+            filter_lang=filter_lang,
+            sortby=sortby,
+            fields=fields,
+            q=q,
+            modifier=modifier,
+        )
+
+        if client and client._stac_io is not None and stac_io is None:
+            self._stac_io = client._stac_io
+            self._collection_search_extension_enabled = client.conforms_to(
+                ConformanceClasses.COLLECTION_SEARCH
+            )
+            self._collection_search_free_text_enabled = client.conforms_to(
+                ConformanceClasses.COLLECTION_SEARCH_FREE_TEXT
+            )
+            if not self._collection_search_extension_enabled:
+                warnings.warn(
+                    str(DoesNotConformTo("COLLECTION_SEARCH"))
+                    + ". Filtering will be performed client-side where only bbox, "
+                    "datetime, and q arguments are supported"
+                )
+                self._validate_client_side_args()
+            else:
+                if not self._collection_search_free_text_enabled:
+                    warnings.warn(
+                        str(DoesNotConformTo("COLLECTION_SEARCH#FREE_TEXT"))
+                        + ". Free-text search is not enabled for collection search"
+                        "Free-text filters will be applied client-side."
+                    )
+
+        else:
+            self._stac_io = stac_io or StacApiIO()
+            self._collection_search_extension_enabled = (
+                collection_search_extension_enabled
+            )
+            self._collection_search_free_text_enabled = (
+                collection_search_free_text_enabled
+            )
+            self._validate_client_side_args()
+
+    def _validate_client_side_args(self) -> None:
+        """Client-side filtering only supports the bbox, datetime, and q parameters."""
+        args = {key for key, value in self._parameters.items() if value is not None}
+        extras = args - {"bbox", "datetime", "q", "limit", "max_items"}
+
+        if extras:
+            raise ValueError(
+                "Only the limit, max_collections, bbox, datetime, and q arguments are "
+                "supported for client-side filtering but these extra arguments were "
+                "provided: " + ",".join(extras)
+            )
+
+    @lru_cache(1)
+    def matched(self) -> Optional[int]:
+        """Return number matched for search
+
+        Returns the value from the `numberMatched` or `context.matched` field.
+        Not all APIs will support counts in which case a warning will be issued
+
+        Returns:
+            int: Total count of matched collections. If counts are not supported `None`
+            is returned.
+        """
+        params = {**self.get_parameters(), "limit": 1}
```

Comment:
What's your thoughts on just starting paging, instead of doing an extra limit=1 request at the start? The first page would be bigger, but it's one fewer request all-in-all?

```python
iter = self.pages_as_dicts()
page = next(iter, None)
if page is None:
    return None
else if ...:
    # do all the other page checking stuff and return if you got it
count = len(page["collections"])
for page in iter:
    count += len(page["collections"])
return count
```

Reply:
Great idea, I added that change in https://github.com/stac-utils/pystac-client/pull/735/commits/fce4bd03e568a69da871a1f78fa576aecd1e32f9